src/java/org/apache/cassandra/db/compaction/CompactionStrategyFactory.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.compaction;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Map;
+
+import javax.annotation.Nullable;
+
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.schema.CompactionParams;
+
+/**
+ * The factory for compaction strategies and their containers.
+ */
+public class CompactionStrategyFactory
+{
+    private final ColumnFamilyStore cfs;
+    private final CompactionLogger compactionLogger;
+
+    public CompactionStrategyFactory(ColumnFamilyStore cfs)
+    {
+        this.cfs = cfs;
+        this.compactionLogger = new CompactionLogger(cfs.metadata());
+    }
+
+    /**
+     * Reload the existing strategy container, possibly creating a new one if required.
+     *
+     * @param current the current strategy container, or {@code null} if this is the first time we're loading a
+     *                compaction strategy
+     * @param compactionParams the new compaction parameters
+     * @param reason the reason for reloading
+     *
+     * @return Either a new strategy container or the current one, but reloaded with the given compaction parameters.
+     */
+    public CompactionStrategyContainer reload(@Nullable CompactionStrategyContainer current,
+                                              CompactionParams compactionParams,
+                                              CompactionStrategyContainer.ReloadReason reason)
+    {
+        // If we were called due to a metadata change but the compaction parameters are the same then
+        // don't reload since we risk overriding parameters set via JMX
+        if (current != null && !current.shouldReload(compactionParams, reason))
+            return current;
+
+        CompactionStrategyContainer ret;
+
+        // 1. If UCS is being requested, recreate the container, potentially inheriting the enabled state of the
+        // current container.
+        if (compactionParams.klass() == UnifiedCompactionStrategy.class)
```

Comment:
The first attempt to fix this is in fe00b71deff972aa535b19c6e1eaa269f7dac97d. We can probably do a bit better than reflection through static methods - I'll think of something in the beginning of next week.

Reply:
Looks great.

As a nit, I'd consider using a static field rather than method for getting the container class from the strategy to make it a little clearer that it shouldn't change.

Reply:
Fixed in bae266032da8789457ed63f0d4046476f38b3bc6.